zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1221,20 +1237,109 @@ private void reconnect() {
     }
   }
 
+  private class SyncCallbackHandler extends ZkAsyncCallbacks.DefaultCallback implements AsyncCallback.VoidCallback {
+    private String _sessionId;
+
+    SyncCallbackHandler(String sessionId) {
+      _sessionId = sessionId;
+    }
+
+    @Override
+    public void processResult(int rc, String path, Object ctx) {
+      LOG.info("sycnOnNewSession with sessionID {} async return code: {}", _sessionId, rc);
+      callback(rc, path, ctx);
+    }
+
+    @Override
+    public void handle() {
+      // Make compiler happy, not used.
+    }
+
+    @Override
+    protected boolean needRetry(int rc) {
+      try {
+        switch (KeeperException.Code.get(rc)) {
+          /** Connection to the server has been lost */
+          case CONNECTIONLOSS:
+            return true;
+          default:
+            return false;
+        }
+      } catch (ClassCastException | NullPointerException ex) {
+        LOG.error("Session {} failed to handle unknown return code {}. Skip retrying. ex {}",
+            _sessionId, rc, ex);
+        return false;
+      }
+    }
+  }
+
+  private void doAsyncSync(final ZooKeeper zk, final String path, final long startT,
+      final SyncCallbackHandler cb) {
+    zk.sync(path, cb,
+        new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, false) {
```

Comment:
Is sync counted as write operations? Just to confirm.
I know it is kind of happening in order like a write, but does it actually update anything on the server-side? Note if we record the type differently, then the server-side metric and client-side metric will have different numbers.

Reply:
Sync() does not update any data in the server. However it blocks any read arriving later than it. 

So your point is that if server metrics does not report it as write, nor should we do it here. Otherwise, both should report it as write?

Reply:
As long as they are treated the same on both sides, I'm good.
But it seems to be a read op.

Reply:
> But it seems to be a read op.

Do you mean the server side metrics treat sync() as read, so this one here should be treated as read? Put it another way, we don't record it? 

What is your proposal?

Reply:
As I said, "As long as they are treated the same on both sides, I'm good."
Note this is not a major concern for the opensource world. Because the official Zookeeper server does not treat read/write differently. It only applies to the modified ZK version. Could you please double-check and confirm we are recording them with the same type?

Reply:
Just checked, internal ZK metrics, sync() is treated as read. 
will change to `new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, true)`